e3sm_diags/driver/lat_lon_driver.py

```diff
@@ -2,242 +2,427 @@
+            ds_land_sea_mask: xr.Dataset = test_ds._get_land_sea_mask(season)
+
+            ds_test = test_ds.get_climo_dataset(var_key, season)  # type: ignore
+
+            # If the reference climatology dataset cannot be retrieved
+            # it will be set the to the test climatology dataset which means
+            # analysis is only performed on the test dataset.
             try:
-                mv2 = ref_data.get_climo_variable(var, season)
+                ds_ref = ref_ds.get_climo_dataset(var_key, season)  # type: ignore
+                parameter.model_only = False
             except (RuntimeError, IOError):
-                mv2 = mv1
-                logger.info("Can not process reference data, analyse test data only")
-
+                ds_ref = ds_test
                 parameter.model_only = True
 
-            parameter.viewer_descr[var] = (
-                mv1.long_name
-                if hasattr(mv1, "long_name")
-                else "No long_name attr in test data."
-            )
+                logger.info("Cannot process reference data, analyzing test data only.")
+
+            # Store the variable's DataArray objects for reuse.
+            dv_test = ds_test[var_key]
+            dv_ref = ds_ref[var_key]
+
+            is_vars_3d = has_z_axis(dv_test) and has_z_axis(dv_ref)
+            is_dims_diff = has_z_axis(dv_test) != has_z_axis(dv_ref)
 
-            # For variables with a z-axis.
-            if mv1.getLevel() and mv2.getLevel():
+            if not is_vars_3d:
+                for region in regions:
+                    _get_metrics_by_region(
+                        parameter,
+                        ds_test,
+                        ds_ref,
+                        var_key,
+                        region,
+                        ref_name,
+                        season,
+                        ds_land_sea_mask,
+                    )
+            elif is_vars_3d:
                 plev = parameter.plevs
-                logger.info("Selected pressure level: {}".format(plev))
+                logger.info("Selected pressure level(s): {}".format(plev))
 
-                mv1_p = utils.general.convert_to_pressure_levels(
-                    mv1, plev, test_data, var, season
-                )
-                mv2_p = utils.general.convert_to_pressure_levels(
-                    mv2, plev, ref_data, var, season
-                )
+                ds_test = regrid_z_axis_to_plevs(ds_test, var_key, parameter.plevs)
+                ds_ref = regrid_z_axis_to_plevs(ds_ref, var_key, parameter.plevs)
 
-                # Select plev.
-                for ilev in range(len(plev)):
-                    mv1 = mv1_p[
-                        ilev,
-                    ]
-                    mv2 = mv2_p[
-                        ilev,
-                    ]
+                for ilev, _ in enumerate(plev):
+                    ds_test_ilev = ds_test[ilev]
+                    ds_ref_ilev = ds_ref[ilev]
```

Comment:
TODO: Fix this xarray subsetting syntax.

This syntax tries to get a variable based on `ilev`, rather than subset on `ilev`.
[`.isel`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.isel.html) instead.

Reply:
Should be done with [`a19e9a4` (#677)](https://github.com/E3SM-Project/e3sm_diags/pull/677/commits/a19e9a4148b8ba18135104aae7ed80cc5a67e02d), just need to test.